src/Control/Exception/Safe.hs

```diff
@@ -172,7 +172,9 @@ evaluateDeep action = do
 catchAnyDeep :: (C.MonadCatch m, MonadIO m, NFData a) => m a -> (SomeException -> m a) -> m a
 catchAnyDeep = catchDeep
 
--- | 'catch' without async exception safety
+-- | 'catch' without async exception safety.
+--
+-- Will catch extension types which are children of 'Control.Exception.SomeAsyncException', see <https://github.com/fpco/safe-exceptions#determining-sync-vs-async>
```

Comment:
> it will catch _all_ exception types

And that was my original understanding after reading the _existing_ documentation. But as I started using the library, I found out that its behavior it's more complicated than that.

To summarize, there are two ways to throw an exception:

- `throw`/ `throwM`/ `throwIO`, dealing with sync exceptions
- `throwTo` for async exceptions.

And two families of exceptions:

- synchronous, children of `SomeException`
- asynchronous, children of `SomeAsyncException`.

`Control.Exception.catch` from the `base` package makes no difference between the two families, while the whole point of `safe-exceptions` is to handle them differently.

Let's now get to `catchAsync`. It does not catch _all_ exceptions. Rather, it only works for the following two combinations:

- `throw` of a sync exception
- `throwTo` of an async exception.

It does not catch anything in the following cases:

- `throw` of an async exception
- `throwTo` of a sync exception.

As an example of an exception that gets lost, consider:

```haskell
self <- myThreadId
catchAsync
  (throwTo self DivideByZero)
  (\DivideByZero -> putStrLn "received DivideByZero")
```

You may say that the above code makes no sense to begin with, as it improperly mixes sync and async exceptions. I won't argue against that, but the fact remains that this seemingly innocuous code is actually doing something unexpected. Moreover, the type system is not helping at all in pinpointing this problem.

To me, there are two ways to improve this. One would be to explicitly state in the `catchAsync` documentation that mixing sync / async `throw`s and exceptions is not supported.

The other one would be to modify `catchAsync` implementation, to make it more similar to `Control.Exception.catch`. I think that would match the expectation of most users, and would also reduce the number of edge cases that have to be considered.